File: Stacks/asteroid_collision.py
from typing import List


class AsteroidCollision:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            destroyed = False
            while stack and stack[-1] > 0 and asteroid < 0 and not destroyed:
                left_asteroid = stack[-1]
                right_asteroid = -asteroid
                if left_asteroid > right_asteroid:
                    destroyed = True
                elif left_asteroid == right_asteroid:
                    destroyed = True
                    stack.pop()
                else:
                    stack.pop()
            if not destroyed:
                stack.append(asteroid)
        return stack

    # Time complexity: O(n) where n is the number of asteroids in the input list.
    # In the worst case, we may need to process each asteroid once,
    # and each asteroid can be added and removed from the stack at most once.
    # Space complexity: O(n) in the worst case,
    # if all asteroids are moving in the same direction and none of them collide,
    # we would end up storing all asteroids in the stack.

File: Stacks/AsteroidCollision.java
package Stacks;

import java.util.ArrayDeque;
import java.util.Deque;

public class AsteroidCollision {
  public int[] asteroidCollision(int[] asteroids) {
    Deque<Integer> stack = new ArrayDeque<>();
    for (int asteroid : asteroids) {
      boolean isDestroyed = false;
      while (!stack.isEmpty() && stack.peekLast() > 0 && asteroid < 0 && !isDestroyed) {
        int leftAsteroid = stack.peekLast();
        int rightAsteroid = -asteroid;
        if (leftAsteroid > rightAsteroid) {
          isDestroyed = true;
        } else if (leftAsteroid == rightAsteroid) {
          isDestroyed = true;
          stack.pollLast();
        } else {
          stack.pollLast();
        }
      }
      if (!isDestroyed) {
        stack.offerLast(asteroid);
      }
    }

    int stackSize = stack.size();
    int[] ans = new int[stackSize];
    int i = 0;
    for (int asteroid : stack) {
      ans[i++] = asteroid;
    }

    return ans;
  }
}

// Time Complexity O(n) - Each asteroid is processed at most twice
// (once when it is added to the stack and once when it is removed).
// Space Complexity O(n) - In the worst case,
// all asteroids could be moving in the same direction and end up in the stack.